### ISSUE TITLE
Document Gradle task properties

### ISSUE THREAD

**Issue by pledbrook** [boxfuse/boxfuse-issues#34]
The Boxfuse website currently only displays the convention properties exposed by the Gradle plugin. It should also display what properties are available on the tasks themselves because it's common to customise these in a build file.

Here are a couple of reasons why:

* You can't lazily configure tasks through convention properties, changing the values based on other aspects of the build.
* It's awkward to have multiple configurations for a task. I want to specify different environment variables for the `run` task depending on whether it's going to AWS or Virtual Box.

Ideally we'd be able to create our own instances of the tasks and not just rely on the task instances created by the plugin.

**Comment by axelfontaine** [comment]
From what we described in our introductary Gradle blog post (https://boxfuse.com/blog/gradle-plugin.html), what exactly do you feel we should implement differently or is currently missing?

In the blog post we showed how to create new instances of tasks and configure them dynamically. If there is a better way to achieve that, we are certainly open for suggestions :-)

**Comment by pledbrook** [comment]
Interesting. I didn't see that. I'm glad that the features are there, which simply makes this a documentation issue.

As I see it, this information should be available through the [Gradle plugin documentation](https://boxfuse.com/docs/gradle/). I suggest two options:

1. For each task, show the fully qualified class name that links to the API docs for that class
2. Incorporate the fully qualified task class name _and_ all its public properties & methods into the existing doc pages.

I prefer the second approach, but I suspect the first is easier to implement and maintain for you.

**Comment by axelfontaine** [comment]
Good points. I've just added a link to the blog post from our main Gradle documentation page. We'll also be evaluating the two approaches you suggested and implement changes very soon.

**Comment by axelfontaine** [comment]
This has now been fixed.

We added documentation for defining additional instances of the tasks and extension: https://boxfuse.com/docs/gradle/#extension

As well as the fully qualified class name of each task. Ex: https://boxfuse.com/docs/gradle/create.html

Happy to improve this further if you have suggestions.

Cheers
Axel

**Comment by pledbrook** [comment]
If the plugin were open source, I could have provided pull requests ;)

Jesting aside, that is much improved on the documentation front, but I think there's a fundamental flaw in the plugin design. It seems that the all tasks pull their configuration from the extension or from project/system properties. This is a non-standard approach and makes it more difficult to customise the build.

I have to admit, the technique of creating new instances of the extension is novel and does work to a degree. However, I would much prefer to see the more standard approach used. This would involve making all the configuration options of each task available as public properties. None of these would be configured by the task itself, which relies on the plugin to initialise its properties.

As an example, [here's the Codenarc task](https://github.com/gradle/gradle/blob/master/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CodeNarc.groovy) that's part of the Gradle distribution. Note how none of its inputs are initialised by the task itself. It basically relies on configuration injection.

So how are its values set from the plugin's convention DSL? Well, the convention DSL is implemented as [an extension](https://github.com/gradle/gradle/blob/master/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CodeNarcExtension.groovy) (just like Boxfuse's). This is largely a dumb data class. The configuration is handled [within the plugin](https://github.com/gradle/gradle/blob/master/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CodeNarcPlugin.groovy#L56), which maps the extension properties to the task's.

To be fair, this isn't documented anywhere as far as I know, and the `conventionMapping` property is an internal API (although widely used). The key principle, though, whatever approach is taken is that the task types should expose their configuration properties.

**Comment by axelfontaine** [comment]
Fair enough about open sourcing the plugin. We actually want to do that and are actively considering it, both for the Maven and the Gradle plugin. We just aren't sure yet how and when to best publish our current internal core API :-)

I fully agree with you that we should stick to path of least surprise. To be honest we searched the internet and just couldn't find anything like a "Gradle plugin implementation best practices guide" and we figured the extension approach would work.

I am a bit wary about using conventionMapping though: http://stackoverflow.com/questions/25379967/retrieving-conventionmapping-in-custom-plugin

Would you mind posting an example of how you feel the "ideal" build.gradle would look like when using Boxfuse? This would be a great help and guideline when implementing this.

**Comment by pledbrook** [comment]
Yes, there's not much information on plugin best practices. It's something I'd like to create at some point.

Your concern about `conventionMapping` is valid, but at this point there is no standard alternative. Until there is one (which probably [won't happen until Gradle 3](https://twitter.com/cedricchampeau/status/636097614476390400), `conventionMapping` will remain as the Gradle team can't afford to remove it due to widespread use.

In other words, whatever you do, it will probably need changing when the new rule model is finally in place. And at least there are plenty of examples of using `conventionMapping`.

Hope that helps!

**Comment by axelfontaine** [comment]
We have just released Boxfuse Client 1.9.6. You can now configure all properties directly on the task in addition to the existing options. Example added to the docs: https://boxfuse.com/docs/gradle/#multiple

**Comment by pledbrook** [comment]
Great news, thanks! Are there Javadocs/Groovydocs for the tasks? If not, I can use the IDE to work it out I think.

**Comment by axelfontaine** [comment]
The next release will include a javadoc jar. For now you can refer to the properties described on the website.
